### PR TITLE
fix(expect): improve the error message when nothing is thrown when ing `toThrow`

### PR DESCRIPTION
**Pull request by Dunqing**
### Description

fix: #3941 
image

### Please don't delete this checklist! Before submitting the PR, please make sure you do the following:
- [x] It's really useful if your PR references an issue where it is discussed ahead of time. If the feature is substantial or introduces breaking changes without a discussion, PR might be closed.
- [x] Ideally, include a test that fails without this PR but passes with it.
- [x] Please, don't make changes to `pnpm-lock.yaml` unless you introduce a new test example.

### Tests
- [x] Run the tests with `pnpm test:ci`.

### Documentation
- [x] If you introduce new functionality, document it. You can run documentation with `pnpm run docs` command.

### Changesets
- [x] Changes in changelog are generated from PR name. Please, make sure that it explains your changes in an understandable manner. Please, prefix changeset messages with `feat:`, `fix:`, `perf:`, `docs:`, or `chore:`.

### CODE REVIEW

**Review by sheremet-va — CHANGES_REQUESTED** [review]
You checked that you added a test, but I don't see any tests.

**Comment by Dunqing** [comment]
> You checked a checkbox that you added a test, but I don't see any tests.

Sorry, test case added

File: packages/expect/src/jest-expect.ts [hunk 1]
```diff
@@ -519,12 +519,22 @@ export const JestChaiExpect: ChaiPlugin = (chai, utils) => {
...
       try {
         obj()
       }
       catch (err) {
+        isThrow = true
         thrown = err
       }
+
+      if (!isThrow && !isNot) {
+        const message = utils.flag(this, 'message') || 'expected function to throw an error, but it didn`t'
```
Comment (sheremet-va) [inline packages/expect/src/jest-expect.ts:532]:
```suggestion
        const message = utils.flag(this, 'message') || 'expected function to throw an error, but it didn\'t'
```

File: test/core/test/jest-expect.test.ts [hunk 2]
```diff
@@ -332,6 +332,13 @@ describe('jest-expect', () => {
     expect(null).toBe(true)
   })
 
+  it.fails('toThrow should failure when does not throw ', () => {
+    expect(() => {
+    }).toThrow(Error)
+    expect(async () => {
```
Comment (sheremet-va) [inline test/core/test/jest-expect.test.ts:338]: This line will never run. Either make 2 separate tests with `.fails` or wrap it inside another `expect(() => {}).toThrow()` and remove `.fails`

**Review by sheremet-va — CHANGES_REQUESTED** [review]

File: test/core/test/jest-expect.test.ts [hunk 3]
```diff
@@ -375,6 +368,18 @@ describe('jest-expect', () => {
       },
     ])
   })
+
+  it('toThrow didn\'t throw', () => {
+    expect(() => {
+      expect(async () => {
+      }).toThrow(Error)
```
Comment (sheremet-va) [inline test/core/test/jest-expect.test.ts:375]: Async error should be with `.rejects`:

```ts
await expect(async () => {}).rejects.toThrow(Error)
```

**Review by sheremet-va — CHANGES_REQUESTED** [review]

File: test/core/test/jest-expect.test.ts (hunk 3 above)
Comment (Dunqing) [reply test/core/test/jest-expect.test.ts:375]: The results are consistent with `jest`
image

Comment (sheremet-va) [reply test/core/test/jest-expect.test.ts:375]: Yes, but it doesn't mean that code is correct. Function didn't finish running yet

Comment (sheremet-va) [reply test/core/test/jest-expect.test.ts:375]: `toThrow` receives a Promise, it doesn't wait it to resolve to _ACTUALLY_ check that error didn't happen

If you throw an error there, you will get unhandled promise error, not `toThrow` error

Comment (Dunqing) [reply test/core/test/jest-expect.test.ts:375]: This test case actually comes from [here](https://github.com/vitest-dev/vitest/issues/3941), which is exactly what I wanted to test

Comment (Dunqing) [reply test/core/test/jest-expect.test.ts:375]: > Async error should be with `.rejects`:
> 
> ```ts
> await expect(async () => {}).rejects.toThrow(Error)
> ```

I got an error
image

Comment (sheremet-va) [reply test/core/test/jest-expect.test.ts:375]: > This test case actually comes from https://github.com/vitest-dev/vitest/issues/3941, which is exactly what I wanted to test

Well, then your error description is not enough "but it didn't threw" doesn't really tell much. In the current example there are two errors: 1) async wasn't awaited, 2) error wasn't thrown

> I got an error

Please, read how to test asynchronous code: https://vitest.dev/api/expect.html#resolves

**Pull request merged**